### ISSUE TITLE
[FEATURE] Run integration tests against ARM64

### ISSUE THREAD

**Issue by Kern--** [awslabs/soci-snapshotter#394]
**Is your feature request related to a problem? Please describe.**
Currently, our automated tests only run on x86_64 machines. We should also run tests against ARM64.

**Describe the solution you'd like**
A matrix of architectures on which to build and test SOCI in our github actions yaml file. We can use something like https://github.com/uraimo/run-on-arch-action to provide the alternate architectures via QEMU. 

**Describe alternatives you've considered**

1) We could alternatively use self-hosted runners with native arch support. This might be limiting in the future since only X86_64, i386, and ARM64 are supported today. https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#architectures

2) We could write our own qemu wrapper script to build/run to make it easier to run against other architectures locally.

**Additional context**
N/A

**Comment by sparr** [comment]
Efforts on #620 revealed this to be a less straightforward ask than expected, due to a bug in containerd and docker related to running arm containers on x86 hosts. This issue is on hold until containerd and docker built with [this patch](https://github.com/containerd/containerd/pull/7636) are available packaged for Ubuntu. There is an [open PR](https://github.com/containerd/containerd/pull/8533) to backport that fix to containerd 1.6 which would lead to its availability sooner, but either way this could take months.

**Comment by sparr** [comment]
https://docs.docker.com/engine/release-notes/24.0/

Docker Engine 24 is built using containerd 1.7.1 which should unblock this issue.

**Comment by sondavidb** [comment]
Per #847, efforts for this are still not looking great. The release of containerd v1.7 came with [this fix](https://github.com/containerd/containerd/pull/7636) for the arch detection bug on emulated software, so we got a lot further. However, it seems that multiple places in the log still show problems with proper emulation detection (see logs in #847).

@turan18 and I looked further into it, and it seems that all of the tests that fail attempt to pull from the dummy registry created in these tests. One example is `[TestLazyPullWithSparseIndex](https://github.com/awslabs/soci-snapshotter/blob/7cc04aff01bce06f3b1759cf1eecf4cebdeeb2cb/integration/pull_test.go#L158)`. Here, it runs `[copyImage](https://github.com/awslabs/soci-snapshotter/blob/eebd84e352a64731a92412447c6dbfcd198328cd/integration/util_test.go#L342)`, which uses `nerdctl push` and pushes with the correct platform (linux/arm/v8). However, when attempting to confirm that it is in the registry via `[getManifestDigest](https://github.com/awslabs/soci-snapshotter/blob/eebd84e352a64731a92412447c6dbfcd198328cd/integration/util_test.go#L592)`, it looks for a sha hash pertaining to the amd64 version of the image, NOT the arm64 version. And because it was pushed with a specific platform, the other image manifests for other archs got truncated, so it fails. Specifically, it is `ctr image list` that fails because it cannot find an image pertaining to the correct platform. This is because containerd is asking for the architecture and it believes it is x86_64, which is not what we want.

Further inspection of the logs shows that the Docker Engine version used to run the container is showing x86 architecture (update here to show link to the logs showing this). I think it passes off the detection of the architecture to [containerd](https://github.com/containerd/containerd/blob/140ecc9247386d3be21616fe285021c081f4ea08/platforms/platforms.go) (separate from the containerd version being used to run the tests). The get.docker.com script we use installs, among its dependencies, containerd.io, which is not yet updated to v1.7.0 of containerd.

My understanding of this is as follows: the tests get the wrong architecture because containerd doesn't properly detect the architecture, because it relies on Docker to do so, which has it wrong because of the containerd it uses to install to begin with (Reread that a couple times if you need to :p). I thought that containerd then relied on the GOARCH variable to detect this, but quick testing showed I was incorrect in this assumption (unless I tested improperly).

Workarounds can be seen in my workflow file changes [here](https://github.com/sondavidb/soci-snapshotter/blob/1ee56009c9120e22342c60fe58ae62d1abe21e1a/.github/workflows/build.yml#L79) in #847, where I tried to modify the Docker install script to force ARM64 detection and force download the ARM64 versions of the packages, but to no avail.

Our options from here are as follows:

- Wait for v1.7 of containerd.io (preferred, but not in our control)
- Somehow trick Docker into detecting ARM architecture (probably not preferred but if we need ARM testing faster, probably better)

Sorry for the long comment. Hopefully this analysis is helpful. Please look at the logs in the checks made by #847 for full logs.

**Comment by Kern--** [comment]
This is probably not worth doing until github actions supports ARM runners.

**Comment by Kern--** [comment]
AWS CodeBuild recently added support for self hosted GHA runners: https://docs.aws.amazon.com/codebuild/latest/userguide/action-runner.html. This is probably a good option for automated ARM testing.

**Comment by Kern--** [comment]
Closed via #1310
